src/nbody/barnes_hut.py

```diff
@@ -0,0 +1,373 @@
+            noderrr,
+        ]  # flatten list
+        node.children = [
+            n for n in node.children if len(n.points) > 0
+        ]  # cull empty nodes
+        return node.children
+
+    def recursive_split(self, root: Node) -> None:
+        """
+        Recursively split the root node until the tree is completely generated.
+
+        Arguments:
+            root (Node): The dyadic cube containing all points in the system, serving as the base of the tree.
+        """
+        stack = [root]
+        while stack:
+            node = stack.pop()
+            stack += self.split(node)
+
+    def __init__(
+        self,
+        points: list[Body],
+        max_depth: int | None = None,
+        min_length: float | None = None,
+    ):
+        """
+        Arguments:
+            points (list[Body]): The bodies in the system to be subdivided.
+
+            max_depth (int, optional): The maximum depth before terminating node splitting. May be infinity.
+            min_length (float, optional): The minimum cube sidelength before terminating node splitting. May be zero.
+        """
+        if max_depth is not None:
+            self.max_depth = max_depth
+        if min_length is not None:
+            self.min_length = min_length
+
+        xvals = [p.position[0] for p in points]
+        yvals = [p.position[1] for p in points]
+        zvals = [p.position[2] for p in points]
+
+        xmin = min(xvals)
+        xmax = max(xvals)
+        ymin = min(yvals)
+        ymax = max(yvals)
+        zmin = min(zvals)
+        zmax = max(zvals)
+
+        base_length = 2 * max(
+            xmax - xmin, ymax - ymin, zmax - zmin
+        )  # double length to avoid issues with the half interval convention
+        self.root = self.Node(
+            coord=np.array([xmin, ymin, zmin]),
+            length=base_length,
+            points=points,
+            parent=None,
+        )
+        self.recursive_split(self.root)
+        self.root.compute_masses()
+
+    def compute_accel(self, other: Body, theta: float = 1, G: float = 6.6743e-11):
+        """
+        Compute the acceleration felt by a given Body object from the system using the Barnes-Hut algorithm.
+
+        Arguments:
+            other (Body): The Body experiencing acceleration.
+            theta (float, optional): An accuracy parameter used by the Barnes-Hut grouping algorithm. Defaults to 1.
+            G (float, optional): The gravitational constant. Defaults to 6.6743e-11.
+
+        Returns:
+            np.ndarray: The acceleration vector.
+        """
+        accel = np.zeros(3)
+
+        stack = [self.root]
+
+        while stack:
+            node = stack.pop()
+            dist = np.linalg.norm(node.center_of_mass - other.position)
+            # accept any cube with l/D < theta
+            length = node.length
+            if (
+                dist == 0
+            ):  # if a cube happens to have center of mass at the point, examine children, unless
```

Comment:
```suggestion
            if np.isclose(dist, 0): # if a cube happens to have center of mass at the point, examine children, unless
```
might be more safe for this application, though I don't know exactly what Barnes-hut is doing here and whether things might blow up -> 0

Reply:
`isclose` is probably better to avoid computing a force from a point cell (it's not an issue otherwise, since the algorithm continues down the tree). I think in an ideal world we compute `atol` based on the `MAX_DEPTH` and `MIN_LENGTH` parameters to avoid considering things in different cubes as the same, but the default is fine for now.